File: src/main/java/dsr/amm/homebudget/controller/AccountController.java
package dsr.amm.homebudget.controller;

import dsr.amm.homebudget.controller.exception.ApiException;
import dsr.amm.homebudget.data.dto.*;
import dsr.amm.homebudget.service.AccountService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.List;

import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.web.bind.annotation.RequestMethod.*;

/**
 * Created by knekrasov on 10/15/2018.
 */
@RestController
@RequestMapping("/accounts")
public class AccountController {
    @Autowired
    private AccountService accountService;

    @RequestMapping(value = "/my", method = GET)
    public List<AccountDTO> getCurrent() {
        return accountService.getMyAccounts();
    }

    @RequestMapping(method = POST)
    @ResponseStatus(CREATED)
    public AccountDTO createAccount(@Valid @RequestBody AccountNewDTO newAcc) {
        return accountService.create(newAcc);
    }

    @RequestMapping(value = "/{id}", method = GET)
    public AccountDTO getAccount(@PathVariable("id") Long id) {
        return accountService.getAccountById(id);
    }

    @RequestMapping(value = "/{id}/transactions", method = POST)
    @ResponseStatus(CREATED)
    public TransactionDTO addTransaction(@RequestBody @Valid TransactionDTO tx, @PathVariable("id") Long accountId) {
        if (tx instanceof DepositTxDTO) {
            return accountService.deposit(accountId, (DepositTxDTO) tx);
        }
        else if (tx instanceof WithdrawalTxDTO) {
            return accountService.withdraw(accountId, (WithdrawalTxDTO) tx);
        }
        throw new ApiException("Unsupported transaction type submitted");
    }

    @RequestMapping(value = "/{id}/transactions", method = GET)
    public Page<TransactionDTO> getTransactions(Pageable pageable, @PathVariable("id") Long accountId) {
        return accountService.getAccountTransactions(pageable, accountId);
    }

    @RequestMapping(value = "/{accountId}/transactions/{transactionId}", method = DELETE)
    public void deleteTransaction(
            @PathVariable("accountId") Long accountId,
            @PathVariable("transactionId") Long transactionId) {
        accountService.deleteTransaction(accountId, transactionId);
    }

}

File: src/main/java/dsr/amm/homebudget/controller/AccountController.kt
package dsr.amm.homebudget.controller

import dsr.amm.homebudget.data.dto.*
import dsr.amm.homebudget.service.AccountService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.web.bind.annotation.*

import javax.validation.Valid

import org.springframework.http.HttpStatus.CREATED
import org.springframework.web.bind.annotation.RequestMethod.GET
import org.springframework.web.bind.annotation.RequestMethod.DELETE
import org.springframework.web.bind.annotation.RequestMethod.POST
import java.time.OffsetDateTime
import java.util.*

/**
 * Created by knekrasov on 10/15/2018.
 */
@RestController
@RequestMapping("/account")
open class AccountController @Autowired constructor(open val accountService: AccountService) {

    val current: List<AccountDTO>
        @RequestMapping(method = [GET])
        get() = accountService.myAccounts

    @RequestMapping(method = [POST])
    @ResponseStatus(CREATED)
    fun createAccount(@Valid @RequestBody newAcc: AccountNewDTO): AccountDTO = accountService.create(newAcc)

    @RequestMapping(value = ["/{id}/transactions"], method = [POST])
    @ResponseStatus(CREATED)
    fun addTransaction(
            @RequestBody @Valid tx: TransactionDTO,
            @PathVariable("id") accountId: Long
    ): TransactionDTO? =
            accountService.transaction(accountId, tx)

    @RequestMapping(value = ["/{id}/transactions"], method = [GET])
    fun getTransactions(
            pageable: Pageable,
            @PathVariable("id") accountId: Long,
            @RequestParam("from") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) from: Optional<OffsetDateTime>,
            @RequestParam("to") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) to: Optional<OffsetDateTime>
    ):Page<TransactionDTO> =
            accountService.getAccountTransactions(pageable, accountId, from, to)

    @RequestMapping(value = ["/{accountId}/transactions/{transactionId}"], method = [DELETE])
    fun deleteTransaction(
            @PathVariable("accountId") accountId: Long,
            @PathVariable("transactionId") transactionId: Long
    ): TransactionDTO =
        accountService.deleteTransaction(accountId, transactionId)
}
